src/brand/bhyve/boot

```diff
@@ -309,45 +369,59 @@ for dev in root.findall('./attr[@name]'):
     ppt += 1
 
 # RNG
+
 if opts['rng'] == 'on':
     args.extend(['-s', '{0}:0,virtio-rnd'.format(RNG_SLOT)])
 
 # Extra options
 
-if len(opts['extra']):
+if opts['extra']:
     args.extend(opts['extra'].split(' '))
 
+# Dump configuration
+
+args.extend(['-o', 'config.dump=1'])
+
 # VM name
 
-name = root.get('name')
 args.append(name)
 
-logging.info('Final arguments: {0}'.format(pformat(args)))
-logging.info('{0}'.format(' '.join(args)))
-
-if testmode:
-    sys.exit(0)
-
-if not os.path.exists('/dev/vmmctl'):
-    logging.info('No kernel support for bhyve')
-    sys.exit(0)
-
-if os.path.exists('/dev/vmm/{0}'.format(name)):
-    logging.info('Destroying old bhyve instance')
-    subprocess.run(['/usr/sbin/bhyvectl', '--vm={0}'.format(name), '--destroy'])
-
-while True:
-    logging.info('Starting bhyve')
-    ret = subprocess.run(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-    logging.info("Bhyve exited {0}".format(ret.returncode))
-    logging.error("Error {0}".format(ret.stderr))
-    logging.debug("Output {0}".format(ret.stdout))
-    # 0 - rebooted
-    # 1 - powered off
-    # 2 - halted
-    # 3 - triple fault
-    # 4 - other error
-    if ret.returncode != 0: break
+##############################################################################
+
+logging.debug('Final arguments: {0}'.format(pformat(args)))
+logging.info('{0}'.format(' '.join(map(
+    lambda s: f'"{s}"' if ' ' in s else s, args))))
+
+p = subprocess.run(args, capture_output=True, text=True)
+# config.dump exits with a status code of 1. Other errors indicate a problem.
+if p.returncode != 1:
+    fatal(f'Could not parse configuration: {p.stderr}')
+
+fh = None
+if not testmode:
+    try:
+        fh = tempfile.NamedTemporaryFile(mode='w', dir=f'{zoneroot}/etc',
+            prefix='bhyve.', delete=False)
+    except Exception as e:
+        fatal(f'Could not create temporary file: {e}')
+    else:
+        logging.debug(f'Created temporary file at {fh.name}')
+
+writecfg(fh, '#\n# Generated from zone configuration\n#')
+
+for line in p.stdout.splitlines():
+    if line.startswith('config.dump'): continue
+    writecfg(fh, line)
+
+if not testmode:
+    tf = fh.name
+    fh.close()
```

Comment:
Should we not excplicitly close the fine in test mode too? I know Python closes it at end of execution, but being explicit seems cleaner and safer

Reply:
If we're in test mode, the file was never opened (fh is `None`)